### ISSUE TITLE
High CPU Usage in the presence of Push and many Pending JS Invocations

### ISSUE THREAD

**Issue by muhammetaltindal** [vaadin/flow#17286]
### Description of the bug

Each server roundtrip can lead to high CPU usage if there are many hidden components with pending JS invocations yet to be sent to the browser.

Consider the minimal reproducible example below which contains one button that will show/hide another button once it is being clicked. However there are also hidden cascaded layouts with buttons in it. Note, that once a button is instantiated it will already create a pending JS invocation (see `c.v.f.c.b.Button#initDisableOnClick`)

When you click the button, you can observe that there is a considerable delay until the other button is shown/hidden.

![output](https://github.com/vaadin/flow/assets/1894101/d43973cc-c3b0-45f5-b668-7cf277327f95)

![Untitled](https://github.com/vaadin/flow/assets/1894101/3ae2cd92-8963-43c5-8909-121a6aeef4d2)

### Expected behavior

The performance could be improved here, such that any server roundtrip may not get slower the more invisible components with pending JS invocations there are.

Firstly, you may compute isVisible() for owner components of pending JS invocations once only. Currently there are three places where it is computed again. Two of those places compute it for the full list of pending JS. See `c.v.f.c.i.UIInternals#isDirty` and `c.v.f.c.i.UIInternals#dumpPendingJavaScriptInvocations`.

Secondly, components may only queue JS invocations into the pending list if they are actually going to be visible.

Another idea is related to the usage of the framework itself, so the programmer may try to not instantiate any components that are not going to be shown. Only instantiate those that will be displayed to the user and discard all others.

### Minimal reproducible example

```java
@Push
@Route("")
public class View extends Div {

  public View() {
    for (int i = 0; i < 100; i++) {
      Component layout = deepLayout(0, 100, buttons(100));
      layout.setVisible(false);
      add(layout);
    }
    Button v = new Button("test");
    add(new Button("click", e -> UI.getCurrent().access(() -> v.setVisible(!v.isVisible()))), v);
  }

  private Div buttons(int n) {
    Div div = new Div();
    IntStream.range(0, n).mapToObj(i -> new Button()).forEach(div::add);
    return div;
  }

  private static Component deepLayout(int i, int j, Component child) {
    return i < j ? new Div(deepLayout(++i, j, child)) : child;
  }
}
```

### Versions

- Vaadin / Flow version: 14.10.3
- Java version: 17.0.6
- OS version: macOS

**Comment by muhammetaltindal** [comment]
**Possible improvement** (Iterate only once through the list of pending JS)

Replace `c.v.f.c.i.UIInternals#dumpPendingJavaScriptInvocations` by:
```java
public List<PendingJavaScriptInvocation> dumpPendingJavaScriptInvocations() {
  pendingTitleUpdateCanceler = null;

  if (pendingJsInvocations.isEmpty()) {
    return Collections.emptyList();
  }

  Iterator<PendingJavaScriptInvocation> invocationsIterator = getPendingJavaScriptInvocations().iterator();
  List<PendingJavaScriptInvocation> readyToSend = new ArrayList<>();
  List<PendingJavaScriptInvocation> pendingJs = new ArrayList<>();
  while (invocationsIterator.hasNext()) {
    PendingJavaScriptInvocation invocation = invocationsIterator.next();
    if (invocation.getOwner().isVisible()) {
      invocation.setSentToBrowser();
      readyToSend.add(invocation);
    } else {
      pendingJs.add(invocation);
    }
  }
  pendingJs.forEach(this::registerDetachListenerForPendingInvocation);
  pendingJsInvocations = pendingJs;
  return readyToSend;
}
```

**Comment by muhammetaltindal** [comment]
**Possible improvement** (In the presence of many invisible buttons):

Copy the `c.v.f.c.b.Button` into your project to apply the following changes to it.
- add field `private boolean disableOnClickInitialized;`
- in method `setDisableOnClick` after `if (disableOnClick) {` add
```java
if (!disableOnClickInitialized) {
  initDisableOnClick();
  disableOnClickInitialized = true;
}
```
- delete `onAttach(AttachEvent)` method completely

If you have many invisible buttons that don't have disableOnClick enabled, then this will reduce the amount of pending JS invocations.

**Comment by mcollovati** [comment]
Handling of `disableOnClick` on Button has been improved in the latest versions of Vaadin.
You may ask on the [flow-component repository](https://github.com/vaadin/flow-components/) if there are any plans to back port the changes to Vaadin 14.10

**Comment by mcollovati** [comment]
Vaadin 24 seems not to be affected by the issue, or at least I was unable to replicate it, probably because of the changes to Button in V24.

However, avoiding the double loop when dumping pending invocations sounds like a good suggestion :+1:

**Comment by muhammetaltindal** [comment]
@mcollovati yes, the demo may be adjusted to just call executeJs with something temporary script on the button, this _may_ replicate the issue on Vaadin 24... for example use the following lambda before adding the button to the layout:

```java
i -> {
  Button b = new Button();
  b.executeJs("true");
  return b;
}
```

**Comment by mcollovati** [comment]
Even forcing JS invocations for the buttons, I cannot see CPU spikes nor progress bars showing in V24.
I'll also check on V23

**Comment by mcollovati** [comment]
I tried the sample code on Vaadin 23.3 and 14.10.3, but I can't replicate the issue.
I can see the progress bar for a moment during the loops used to build the layout, but when I click on the first button, the UI update is immediate, no progress bar is shown, and I can't spot any CPU spikes
